File: Jasmine/JasmineCompiler/Src/DFIR_Compiler/DFIR_Compiler/ITunnel.cs
using System;
using System.Collections.Generic;


namespace DFIR_Compiler
{
    public class ITunnel:IBorderNode
    {        
        private bool isInput;
        private ITerminal innerTerminal;
        private ITerminal outerTerminal;
        private IndexingMode indexingMode;

        public ITunnel(int nodeId, int parentId, List<ITerminal> inputTerminals, List<ITerminal> outputTerminals, IndexingMode indexingMode, IsInputMode isInput, int innerTerminalId, int outerTerminalId) : base(nodeId, parentId, inputTerminals, outputTerminals)
        {
            if (isInput == IsInputMode.False)
            {
                this.isInput = false;
            }
            else
            {
                this.isInput = true;
            }
            innerTerminal = ITerminal.GetIdToITerminal()[innerTerminalId];
            outerTerminal = ITerminal.GetIdToITerminal()[outerTerminalId];
            SetIndexingMode(indexingMode);
            INodeType = "ITunnel";
            indegree = 1;
        }

        public virtual bool GetIsInput()
        {
            return isInput;
        }
        public virtual ITerminal GetInnerTerminal()
        {
            return innerTerminal;
        }
        public virtual ITerminal GetOuterTerminal()
        {
            return outerTerminal;
        }
        public virtual IndexingMode GetIndexingMode()
        {
            return indexingMode;
        }
        public override INode GetParentINode()
        {
            return IdToINode[parentId];
        }


        public virtual void SetIndexingMode(IndexingMode indexingMode)
        {
            this.indexingMode = indexingMode;
        }

        public virtual string GetILoopIndexName()
        {
            IForLoop forloop = (IForLoop)GetParentINode();
            return forloop.GetILoopIndex().GetName();
        }
        public virtual string GetIndexingITunnelName()
        {
            return GetName() + '[' + GetILoopIndexName() + ']';
        }


        public virtual bool IsOutputCaseITunnel()
        {
            bool result = false;
            if (GetParentINode().GetINodeType() == "ICaseStructure")
            {
                if (!GetIsInput())
                {
                    result = true;
                }
            }
            return result;
        }

        public virtual bool IsIndexingITunnelInside(IDiagram iDiagram)
        {
            return( (IsBorderNodeInside(iDiagram)) && (indexingMode == IndexingMode.Indexing));
        }

        public override void DecreaseIndegree()
        {
            indegree--;
            if (isInput)
            {
                IdToINode[parentId].indegree--;
            }

        }


        public override void Print()
        {
            Console.WriteLine("------------------------------------");
            Console.WriteLine("ITunnel:");
            Console.WriteLine("NodeId: " + nodeId + ' ' + "ParentId: " + parentId);
            Console.WriteLine("IsInput: " + GetIsInput());
            Console.WriteLine("GetInnerTerminal: " + GetInnerTerminal().GetITerminalId());
            Console.WriteLine("GetOuterTerminal: " + GetOuterTerminal().GetITerminalId());
            Console.WriteLine("IndexingMode: " + GetIndexingMode());
            for (int i = 0; i < inputTerminals.Count; i++)
            {
                inputTerminals[i].Print();
            }
            for (int i = 0; i < outputTerminals.Count; i++)
            {
                outputTerminals[i].Print();
            }
        }
    }
}

File: Wayne/WayneCompiler/src/Dfir_representation/ITunnel.java
package Dfir_representation;

import java.util.Vector;


public class ITunnel extends IBorderNode{
	   
	   private IndexingMode indexingMode;
	   private Boolean isInput;
	   private ITerminal innerTerminal;
	   private ITerminal outerTerminal;
	   
	   public ITunnel(int nodeId, int parentId, Vector<ITerminal> inputTerminals, 
			          Vector<ITerminal> outputTerminals,  IndexingMode indexingMode, 
			          IsInputMode isInput, int innerTerminalId, int outerTerminalId){
		   super(nodeId,parentId,inputTerminals,outputTerminals);
		   if(isInput == IsInputMode.valueOf("False")){
			   this.isInput = false;
		   }
		   else{
			   this.isInput = true;
		   }
		   this.innerTerminal = ITerminal.getIdToITerminal().get(innerTerminalId);
		   this.outerTerminal = ITerminal.getIdToITerminal().get(outerTerminalId);
		   this.indexingMode=indexingMode;
		   JNodeType = "ITunnel";
		   this.indegree = 1;
	   }
	   
	   
	   public Boolean isOutputITunnelForCase(){
		   Boolean res = false;
		   if(this.getParentJNode().getJNodeType() == "ICaseStructure"){
			   if(!this.getIsInput()){
				   res = true;
			   }
		   }
		   return res;
	   }
	   public Boolean isIndexingITunnelInsideLoop(IDiagram iDiagram){
			return this.isBorderNodeInsideLoop(iDiagram) && indexingMode == IndexingMode.valueOf("Indexing");
		}
	   public String getILoopIndexName(){
		   IForLoop f = (IForLoop)this.getParentJNode();
		   return f.getILoopIndex().getName();
	   }
	   public String getIndexingITunnelName(){
		   return this.getName() + '[' + this.getILoopIndexName() + ']';
	   }
	   public JNode getParentJNode(){
		   return JNode.IdToJNode.get(this.parentId);
	   }
	   public void decIndegree(){
		   this.indegree--;
		   if(this.isInput){
			   JNode.IdToJNode.get(this.parentId).indegree--;
		   }
		   
	   }
	   public Boolean getIsInput() {
		   return this.isInput;
	   }
	   public ITerminal getInnerTerminal(){
		   return this.innerTerminal;
	   }
	   public ITerminal getOuterTerminal(){
		   return this.outerTerminal;
	   }
	   public void print(){
		   System.out.println("ITunnel begin");
		   System.out.println("NodeId: " + this.nodeId + ' ' + "ParentId: " + this.parentId);
		   System.out.println("IsInput: " + this.getIsInput());
		   System.out.println("GetInnerTerminal: " + this.getInnerTerminal().getITerminalId());
		   System.out.println("GetOuterTerminal: " + this.getOuterTerminal().getITerminalId());
		   System.out.println("IndexingMode: " + this.getIndexingMode());
		   for(int i=0; i<inputTerminals.size(); i++){
				   inputTerminals.get(i).print();
		   }
		  for(int i=0; i<outputTerminals.size(); i++){
				   outputTerminals.get(i).print();
		   }   
		  System.out.println("ITunnel end\n");
	   }
	   public IndexingMode getIndexingMode() {
		    return indexingMode;
	   }
	   public void setIndexingMode(IndexingMode indexingMode) {
		    this.indexingMode = indexingMode;
	   }
}

